SwiftCompilerSources/Sources/Optimizer/InstructionSimplification/SimplifyPointerToAddress.swift

```diff
@@ -0,0 +1,58 @@
+//===--- SimplifyPointerToAddress.swift -----------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2023 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import SIL
+
+extension PointerToAddressInst : OnoneSimplifyable {
+
+  /// For a redundant pair of pointer-address conversions, e.g.
+  ///
+  ///   %2 = address_to_pointer %1
+  ///   %3 = pointer_to_address %2 [strict]
+  ///
+  /// replace all uses of %3 with %1.
+  ///
+  func simplify(_ context: SimplifyContext) {
+    if let atp = self.pointer as? AddressToPointerInst,
+       atp.address.type == self.type,
+       self.isStrict,
+       // If the pointer is escaped from a borrow-scope, the transformation would break ownership rules.
+       !atp.mayBeEscapedPointerFromBorrowScope
+    {
+      self.uses.replaceAll(with: atp.address, context)
+    }
+  }
+}
+
+private extension AddressToPointerInst {
+
+  /// A simple check if the pointer could be escaped from a borrow scope. E.g.
+  ///
+  ///   %2 = begin_borrow %1
+  ///   %3 = rel_tail_addr %2
+  ///   %4 = address_to_pointer %3
+  ///   end_borrow_%2
+  ///   // use of %4
+  ///
+  var mayBeEscapedPointerFromBorrowScope: Bool {
+    if !parentFunction.hasOwnership {
+      return false
+    }
+    switch address.accessBase {
```

Comment:
This switch looks similar to `AccessBase::hasLocalOwnershipLifetime`, except that it's more conservative for values with none ownership

```
  case AccessBase::Box:
  case AccessBase::Class:
  case AccessBase::Tail:
    return getReference()->getOwnershipKind() != OwnershipKind::None;
```

 and

```
  case AccessBase::Unidentified:
    // Unidentified storage may be nested within object access, but this is an
    // "escaped pointer", so it is not restricted to the object's borrow scope.
    return false;
```

Would it make sense for `mayBeEscapedPointerFromBorrowScope` to call through to `hasLocalOwnershipLifetime`?  Or, if that other function isn't easily bridgeable, for a version of `hasLocalOwnershipLifetime` to be added to AccessBase in the Swift sources?